vllm/v1/sample/tree_rejection_sampler.py

```diff
@@ -0,0 +1,232 @@
+                                              accepted_path_levels,
+                                              accepted_path_indices]
+        path_masks = tree_mask[batch_indices, accepted_paths]
+
+        # Create output buffer.
+        num_req = len(metadata.num_draft_tokens)
+        output_token_ids = torch.empty(
+            (num_req, self.tree_depth),
+            dtype=torch.
+            int32,  # Consistent with SamplerOutput.sampled_token_ids.
+            device=draft_token_ids.device,
+        )
+        output_token_ids.fill_(PLACEHOLDER_TOKEN_ID)
+
+        # Set accepted draft tokens.
+        col_indices = torch.cumsum(path_masks, dim=1) - 1
+        row_indices = batch_indices.unsqueeze(1).expand((-1, draft_tree_size))
+        scatter_rows = row_indices[path_masks]
+        scatter_cols = col_indices[path_masks]
+        accepted_draft_token_ids = draft_token_ids[path_masks]
+        output_token_ids[scatter_rows, scatter_cols] = accepted_draft_token_ids
+
+        # Sample and add a bonus token to the accepted paths.
+        bonus_target_logits = tree_target_logits[batch_indices, accepted_paths]
+        bonus_sampler_output = self.main_sampler(
+            logits=bonus_target_logits,
+            sampling_metadata=sampling_metadata,
+        )
+        output_token_ids[
+            batch_indices,
+            accepted_path_levels] = bonus_sampler_output.sampled_token_ids.view(
+                -1)
+
+        if num_req > num_tree_decodes:
+            # In some cases, we may have leftover requests with 0 draft tokens.
+            # Sample a token for each.
+            bonus_sampler_output = self.main_sampler(
+                logits=target_logits[num_tree_logits:],
+                sampling_metadata=sampling_metadata,
+            )
+            output_token_ids[num_tree_decodes:,
+                             0] = bonus_sampler_output.sampled_token_ids.view(
+                                 -1)
+        return output_token_ids
+
+    def compute_probs(self, logits: torch.Tensor,
+                      sampling_metadata: SamplingMetadata):
+        if sampling_metadata.all_greedy:
+            return logits
+
+        assert sampling_metadata.temperature is not None
+        batch_size, tokens_per_batch, vocab_size = logits.shape
+        logits.div_(sampling_metadata.temperature.view(-1, 1, 1))
+
+        top_k = None
+        if sampling_metadata.top_k is not None:
+            top_k = sampling_metadata.top_k.repeat_interleave(tokens_per_batch)
+        top_p = None
+        if sampling_metadata.top_p is not None:
+            top_p = sampling_metadata.top_p.repeat_interleave(tokens_per_batch)
+        logits = apply_top_k_top_p(logits.view(-1, vocab_size), top_k,
+                                   top_p)
+        output_probs = logits.softmax(dim=-1, dtype=torch.float32)
+        return output_probs.view(batch_size, -1, vocab_size)
+
+    def rejection_sample(self, target_probs: torch.Tensor,
```

Comment:
This looks like greedy acceptance criterion, which will in general have lower acceptance rate than the probabilistic/rejection sampling-style algo used by EAGLE. IIRC chain spec decoding in vLLM supports the probabilistic acceptance. Do you also plan to implement it for trees? You can have a look at Algorithm 1 in https://arxiv.org/pdf/2401.15077.